design/backup-pvc-config.md

```diff
@@ -0,0 +1,94 @@
+- For some storage providers, when creating a read-only volume from a snapshot, it is very fast; whereas, if a writable volume is created from the snapshot, they need to clone the entire disk data, which is time consuming. If the backupPVC's `accessModes` is set as `ReadOnlyMany`, the volume driver is able to tell the storage to create a read-only volume, which may dramatically shorten the snapshot expose time. On the other hand,  `ReadOnlyMany` is not supported by all volumes. Therefore, users should be allowed to configure the `accessModes` for the backupPVC.  
+- Some storage providers create one or more replicas when creating a volume, the number of replicas is defined in the storage class. However, it doesn't make any sense to keep replicas when an intermediate volume used by the backup. Therefore, users should be allowed to configure another storage class specifically used by the backupPVC.  
+
+## Goals
+
+- Create a mechanism for users to specify various configurations for backupPVC    
+
+## Non-Goals
+
+## Solution
+
+We will use the ```node-agent-config``` configMap to host the backupPVC configurations.
+This configMap is not created by Velero, users should create it manually on demand. The configMap should be in the same namespace where Velero is installed. If multiple Velero instances are installed in different namespaces, there should be one configMap in each namespace which applies to node-agent in that namespace only.  
+Node-agent server checks these configurations at startup time and use it to initiate the related Exposer modules. Therefore, users could edit this configMap any time, but in order to make the changes effective, node-agent server needs to be restarted.  
+Inside ```node-agent-config``` configMap we will add one new kind of configuration as the data in the configMap, the name is ```backupPVC```.  
+Users may want to set different backupPVC configurations for different volumes, therefore, we define the configurations as a map and allow users to specific configurations by storage class. Specifically, the key of the map element is the storage class name used by the sourcePVC and the value is the set of configurations for the backupPVC created for the sourcePVC.   
+
+The data structure for ```node-agent-config``` is as below:
+```go
+type Configs struct {
+	// LoadConcurrency is the config for data path load concurrency per node.
+	LoadConcurrency *LoadConcurrency `json:"loadConcurrency,omitempty"`
+
+	// LoadAffinity is the config for data path load affinity.
+	LoadAffinity []*LoadAffinity `json:"loadAffinity,omitempty"`
+
+	// BackupPVC is the config for backupPVC of snapshot data movement.
+	BackupPVC map[string]BackupPVC `json:"backupPVC,omitempty"`
+}
+
+type BackupPVC struct {
+	// StorageClass is the name of storage class to be used by the backupPVC.
+	StorageClass string `json:"storageClass,omitempty"`
+
+	// ReadOnly sets the backupPVC's access mode as read only.
+	ReadOnly bool `json:"readOnly,omitempty"`
```

Comment:
I think this is probably fine as-is, but I'm wondering whether it might be more flexible to specify the access mode explicitly rather than as a read-only bool -- i.e. make it a string and have the user specify ReadOnlyMany, ReadWriteOnce, or ReadWriteMany -- if blank, we use ReadWriteOnce. One reason to consider this is if there is ever a possible scenario for a user to want to specify RWX. If this will never be needed by any user, then the bool may be more user-friendly.

Reply:
As much as I would prefer backup to be read-only by default, there are storageclasses that do not support ReadOnlyMany. So this does need to be opt-in and the default to keep the existing accessmodes.

Another suggestion, is an option for _alternate_ storageclass on expose as well.

Cephfs, which benefits from this has an issue with a specific version. The ceph-csi driver v3.7 which first implemented shallow-copy on restore, was disabled by default. Since storageclass is immutable, enabling this feature required a copy of the storageclass with the feature enabled in parameters. 

Other csi drivers have also put snapshot related parameters in the storageclass instead of the volumesnapshotclass parameters.

Reply:
@msfrucht yes -- that's spelled out later in the design. If it's false (which is the default for a bool field that's not specified explicity), then it will be ReadWriteOnce. If we end up going with the string approach as I suggested in the comment above, then we will need to set it to ReadWriteOnce if the field is blank.

Reply:
@sseago @msfrucht 
Technically, we should set the backupPVC as readOnly, as the backupPVC is used to read the snapshot data, Velero never needs to write anything to it.
However, PVC's accessModes don't have something called `readOnlyOnce`, and not all storages support `ReadOnlyMany` (mostly because they don't support `Many` mode). Per my understanding, many storages doesn't support it.
On the other side,  setting backupPVC as readOnly won't make the same effect as Ceph for most storages, or for theses storages non-flatten snapshot is not supported.
Therefore, we still set the default mode as `ReadWriteOnce`, which at least won't fail any backups; once users make sure `ReadOnlyMany` works for their storage, they can change the default value.

As the other access mode -- RWX, I don't see it is needed. As mentioned above, Velero should never write anything to the backupPVC. Therefore, I think keeping it as bool value is enough and user-friendly.

Reply:
`ReadOnlyOnce` would have solved this problem nicely. It is the Many part that those storage providers have issues with. But that's not entirely relevant as it would on Kubernetes storage-sig github.

ReadOnly should be used if supported and available for data safety and to accidently prevent writing scratch files into the PVC. Our organization supports and encourages the use of readOnlyRootFilesystem so the scratch folder location has to be set to a mounted volume.  ReadOnly is a backstop just-in-case of that error.

Reply:
@Lyndon-Li I think that should work for all use cases we're aware of. Default to ReadWriteOnce, but change to ReadOnlyMany if configmap exists, and the user has set `ReadOnly=true` for the storageclass.

Reply:
@msfrucht 

> ReadOnly should be used if supported and available for data safety

The primary target for the readOnly backupPVC to achieve is to tell the storage to keep in non-flatten mode when creating a volume from a snapshot (so that storage space is saved and volume creation is much faster).
We don't cover the data safety in this design. Not writing anything to a source volume which is being backed up is a basic quality of a backup tool, so we don't need to make a constraint from this low level.

On the other hand, I agree that readOnly should always be used to the backupPVC, however, as mentioned above, many storages don't support `ReadOnlyMany`, so we have to set the default value as `ReadWriteOnce`. In another word, if Kubernetes supports a readOnly mode that all storages support, we will definitely use it by default; or if Kubernetes provides a way to check the support of `ReadOnlyMany` against the storage, Velero will have a better manner to automatically switch between `ReadOnlyMany` and `ReadWriteOnce`.

Reply:
@Lyndon-Li I'm aware you can't find out whether a storage system supports a certain access mode externally.

That was and is a big problem still that the csi spec has not been corrected for.

Reply:
@msfrucht @Lyndon-Li I think for the purposes of this design we're planning on avoiding the access mode checks and will rely on the user telling us via the configmap that this storageclass should use readonly mode.

@msfrucht It's not just the CSI spec that's missing this, I think it should really be on the StorageClass itself -- Even pre-CSI, I was frustrated with the fact that there was no way of knowing from the StorageClass resource which access modes were supported. In a prior application I worked on, we had to hard-code a list of known access modes for known storageclasses, but because storageclass names (or even provisioner strings) aren't guaranteed, this list could only be used in an advisory capacity -- warning users if they chose an access mode that we didn't think was supported, but we couldn't be certain.

Reply:
@sseago Do you have any further comments? If not, please help to approve it. Thanks.